flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierUnaligner.java

```diff
@@ -409,5 +416,15 @@ synchronized long getCurrentCheckpointId() {
 		boolean isCheckpointPending() {
 			return numBarriersReceived > 0;
 		}
+
+		private void resetReceivedBarriers() {
+			Arrays.fill(storeNewBuffers, false);
+			numBarriersReceived = 0;
+		}
+
+		private void notifyAbort(CheckpointException exception) throws IOException {
+			long currentCheckpointId = currentReceivedCheckpointId;
+			handler.executeInTaskThread(() -> handler.notifyAbort(currentCheckpointId, exception), "notifyAbort");
+		}
```

Comment:
I'm not sure if this is working properly and even if it is, I'm not sure if it's a good way to solve the problem.

As far as I understand it works like this:
1. we trigger checkpoint from netty thread and enqueue `notifyCheckpoint` in the mailbox
2. we receive channel closed/end of partition event before `notifyCheckpoint` starts executing in the task thread from mailbox. This will lead us to this method and enqueuing `notifyAbort` after `notifyCheckpoint` in the mailbox
3. we are still going to execute `notifyCheckpoint` callback and trigger the checkpoint, despite it was already (partially?) aborted (for example `resetReceivedBarriers()` has already been called.
4.  `notifyAbort` will clean up the checkpoint started in 3.

Besides being quite complicated and hard to reason about, I'm not sure if it's correct and what could be the side effects of doing this in so many stages. It also might be unnecessarily using resources for starting a checkpoint that we already know will not happen.

Why can not it work like that:
1. as before
2. we receive channel closed/end of partition event before `notifyCheckpoint` starts executing in the task thread from mailbox. We mark the correct checkpoints as cancelled (by bumping the cancelled/current checkpoint ids) in the `ThreadSafeUnaligner` and we abort the checkpoint immediately (assuming we are in the task thead, but I think we are always).
3. if `notifyCheckpoint` starts executing it should check if the checkpoint it's suppose to notify wasn't cancelled.

Reply:
Yeah, I get your point here. 

`notifyAbort` actually be used in two places ATM, one case is that the netty thread firstly receive the ch1 from one channel and triggered it into mailbox, then it receives the ch2 from another input channel and should abort the previous ch1. But we do not know whether the previous triggered cha1 was executing or not. If not executed, it would be skipped via comparing the current checkpoint id. Otherwise it would be aborted during execution.

Another case is for processing end of partition (eof) as you said above. My previous thought was that in the process of async checkpoint, the task thread can still process input (eof). If so, we can not always assume the previous triggered checkpoint has not started when processing eof.

If my assumption is wrong, your suggestion should be an improvement in this case. But it might have some concerns to couple eof with current checkpoint ids. Because the current semantic of eof only aborts the current pending checkpoint, that means it can still trigger new checkpoint afterwards. So it might destroy the normal checkpoint/cancellation id judgment for the following checkpoints even though some channels are already ended.

E.g. when the eof is processed and the current checkpoint id is 3, we should abort the current ch3. But when we receive ch4 from one channel, we should also trigger ch4 in normal way from the perspective of `CheckpointBarrierHandler`. If we bump the current checkpoint id to 4  or larger id when processing eof, that would break the following new received ids. Unless we make the assumption that `CheckpointCoordinator` would never trigger new checkpoint after processing eof, but we actually did not have such assumption in the past from `CheckpointBarrierHandler` view.

Reply:
I think adding a guard in `handler.executeInTaskThread(() -> handler.notifyCheckpoint(barrier), "notifyCheckpoint");` against executing a cancelled checkpoint is always a good idea. But I'm also seeing that it is not enough.

Reply:
Yeah, we already had this guard in `#notifyCheckpoint` method. So are there any concerns for `#notifyAbort` method?